src/xc_integrator/local_work_driver/device/cuda/kernels/uvvars.cu

```diff
@@ -22,50 +47,331 @@ __global__ void eval_uvars_lda_kernel( size_t        ntasks,
   auto& task = tasks_device[ batch_idx ];
 
   const auto npts            = task.npts;
-  const auto nbf             = task.bfn_screening.nbe;
 
-  auto* den_eval_device   = task.den;
+  auto* den_pos_eval_device   = task.den_s;
+  auto* den_neg_eval_device   = task.den_z;
 
-  const auto* basis_eval_device = task.bf;
 
-  const auto* den_basis_prod_device = task.zmat;
+  const int tid = blockIdx.x * blockDim.x + threadIdx.x;
 
-  const int tid_x = blockIdx.x * blockDim.x + threadIdx.x;
-  const int tid_y = blockIdx.y * blockDim.y + threadIdx.y;
 
-  register double den_reg = 0.;
+  if( tid < npts ) {
+    const auto ps = den_pos_eval_device[ tid ];
+    const auto pz = den_neg_eval_device[ tid ];
+    den_pos_eval_device[ tid ] = 0.5*(ps + pz);
+    den_neg_eval_device[ tid ] = 0.5*(ps - pz);
 
-  if( tid_x < nbf and tid_y < npts ) {
+  }
+}
 
-    const double* bf_col   = basis_eval_device     + tid_x*npts;
-    const double* db_col   = den_basis_prod_device + tid_x*npts;
+__global__ void eval_uvars_lda_gks_kernel( size_t        ntasks,
+                                       XCDeviceTask* tasks_device ) {
 
-    den_reg = bf_col[ tid_y ]   * db_col[ tid_y ];
+  const int batch_idx = blockIdx.z;
+  if( batch_idx >= ntasks ) return;
 
-  }
+  auto& task = tasks_device[ batch_idx ];
 
-  // Warp blocks are stored col major
-  constexpr auto warp_size = cuda::warp_size;
-  //constexpr auto max_warps_per_thread_block = cuda::max_warps_per_thread_block;
-  den_reg = cuda::warp_reduce_sum<warp_size>( den_reg );
+  const auto npts            = task.npts;
 
+  auto* den_z_eval_device   = task.den_s;
+  auto* den_s_eval_device   = task.den_z;
+  auto* den_y_eval_device   = task.den_y;
+  auto* den_x_eval_device   = task.den_x;
+  auto* K_z_eval_device     = task.K_z;
+  auto* K_y_eval_device     = task.K_y;
+  auto* K_x_eval_device     = task.K_x;
+  const double dtolsq = 1e-24;  // TODO: make variable
+
+  const int tid = blockIdx.x * blockDim.x + threadIdx.x;
+
+
+  if( tid < npts ) {
+    const auto ps = den_s_eval_device[ tid ];
+    const auto pz = den_z_eval_device[ tid ];
+    const auto py = den_y_eval_device[ tid ];
+    const auto px = den_x_eval_device[ tid ];
+    const auto mtemp = pz*pz + px*px + py*py;
+    double mnorm = 0.;
+  
+    if (mtemp > dtolsq) {
+      mnorm = sqrt(mtemp);
```

Comment:
XXX Better to store 1/mnorm and multiply - division has terrible throughput in general